store/tikv/coprocessor.go

```diff
@@ -275,52 +330,76 @@ func buildCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv
 				respChan:  make(chan *copResponse, 2),
 				cmdType:   cmdType,
 				storeType: req.StoreType,
-			})
+			}
 		}
 	}
-
-	err := splitRanges(bo, cache, ranges, appendTask)
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+	go func() {
+		defer close(receiver)
+		defer close(errCh)
+		if err := splitRanges(bo, cache, ranges, appendTask); err != nil {
+			errCh <- err
+		}
+	}()
 
 	if req.Desc {
+		var tasks []*copTask
+		for task := range receiver {
+			tasks = append(tasks, task)
+		}
 		reverseTasks(tasks)
+		// tasksCh = make(chan *copTask, 1)
+		go func() {
+			defer close(tasksCh)
+			for _, task := range tasks {
+				tasksCh <- task
+			}
+		}()
 	}
-	if elapsed := time.Since(start); elapsed > time.Millisecond*500 {
-		logutil.BgLogger().Warn("buildCopTasks takes too much time",
-			zap.Duration("elapsed", elapsed),
-			zap.Int("range len", rangesLen),
-			zap.Int("task len", len(tasks)))
-	}
-	tikvTxnRegionsNumHistogramWithCoprocessor.Observe(float64(len(tasks)))
-	return tasks, nil
+	return
 }
 
-func buildTiDBMemCopTasks(ranges *copRanges, req *kv.Request) ([]*copTask, error) {
```

Comment:
We can keep `buildTiDBMemCopTasks` as before to make code simple since it's not a problem currently.

Reply:
Yeah, `buildTiDBMemCopTasks` is still there. See:

https://github.com/pingcap/tidb/pull/14384/files#diff-c27388ffb48c6f6eaeefe07dd3243530R380

With the same signature.

Reply:
Got it. Removed `buildCopTasksChan`. Use `buildCopTasks` in a separate goroutine.